services/motion/explore/explore.go

```diff
@@ -0,0 +1,620 @@
+
+	// Start executing plan
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ms.executePlan(cancelCtx, kb, plan)
+	}, ms.backgroundWorkers.Done)
+
+	for {
+		// this ensures that if the context is cancelled we always return early at the top of the loop
+		if err := ctx.Err(); err != nil {
+			return false, err
+		}
+
+		select {
+		// if context was cancelled by the calling function, error out
+		case <-ctx.Done():
+			return false, ctx.Err()
+
+		// once execution responds: return the result to the caller
+		case resp := <-ms.executionResponseChan:
+			ms.logger.Debugf("execution completed: %s", resp)
+			return resp.success, resp.err
+
+		// if the checkPartialPlan process hit an error return it, otherwise exit
+		case resp := <-ms.obstacleResponseChan:
+			ms.logger.Debugf("obstacle response: %s", resp)
+			if resp.err != nil {
+				return resp.success, resp.err
+			}
+			if resp.success {
+				return resp.success, nil
+			}
+		}
+	}
+}
+
+type moveResponse struct {
+	err     error
+	success bool
+}
+
+// checkForObstacles will continuously monitor the generated transient worldState for obstacles in the given
+// motionplan plan. A response will be sent through the channel if an error occurs, the motionplan plan
+// completes or an obstacle is detected in the given range.
+func (ms *explore) checkForObstacles(
+	ctx context.Context,
+	obstacleDetectors []obstacleDetectorPair,
+	kb kinematicbase.KinematicBase,
+	plan motionplan.Plan,
+	extra map[string]interface{},
+) {
+	// Get obstacle polling frequency from extra
+	var obstaclePollingFrequencyHz float64
+	if obstaclePollingFrequencyHzInterface, ok := extra["obstacle_polling_frequency_hz"]; ok {
+		obstaclePollingFrequencyHz, ok = obstaclePollingFrequencyHzInterface.(float64)
+		if !ok {
+			return
+		}
+	} else {
+		obstaclePollingFrequencyHz = defaultObstaclePollingFrequencyHz
+	}
+
+	// Constantly check for obstacles in path at desired obstacle polling frequency
+	ticker := time.NewTicker(time.Duration(int(1000/obstaclePollingFrequencyHz)) * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			currentPose := spatialmath.NewZeroPose()
+
+			// Look for new transient obstacles and add to worldState
+			worldState, err := ms.generateTransientWorldState(ctx, obstacleDetectors)
+			if err != nil {
+				ms.obstacleResponseChan <- moveResponse{err: err}
+				return
+			}
+
+			// Check motionplan plan for transient obstacles
+			collisionPose, err := motionplan.CheckPlan(
+				kb.Kinematics(),
+				plan,
+				worldState,
+				ms.frameSystem,
+				currentPose,
+				[]referenceframe.Input{{Value: 0}, {Value: 0}},
+				spatialmath.NewZeroPose(),
+				ms.logger,
+			)
+			if err != nil {
+				// If an obstacle is detected, check if its within the valid obstacle distance to trigger an
+				// end to the checkForObstacle loop
+				if collisionPose.Point().Distance(currentPose.Point()) < validObstacleDistanceMM {
```

Comment:
[note] Only exit loop and return success: true if an obstacle is detected near the robot (as defined by validObstacleDistanceMM)